gcc/d/d-decls.cc

```diff
@@ -260,10 +280,39 @@ FuncDeclaration::toSymbol()
       // Run full semantic on symbols we need to know about during compilation.
       if (inferRetType && type && !type->nextOf() && !functionSemantic())
 	{
+	  csym = new Symbol();
 	  csym->Stree = error_mark_node;
 	  return csym;
 	}
 
+      // Use same symbol for FuncDeclaration templates with same mangle
+      if (!this->mangleOverride && this->fbody)
+	{
+	  tree ident = get_identifier(mangleExact(this));
+	  if (!IDENTIFIER_DSYMBOL (ident))
+	    {
+	      csym = new Symbol();
+	      IDENTIFIER_DSYMBOL (ident) = this;
+	    }
+	  else
+	    {
+	      bool local_p, template_p;
+	      get_template_storage_info(this, &local_p, &template_p);
+	      // Non-templated functions shouldn't be defined twice
+	      if (!template_p)
+	        {
+		  error("Found duplicated function at %s",
+		    IDENTIFIER_DSYMBOL (ident)->loc.toChars());
```

Comment:
Maybe we could use the same format at duplicate identifiers?  `ScopeDsymbol::multiplyDefined(loc, this, IDENTIFIER_DSYMBOL (ident))` ?